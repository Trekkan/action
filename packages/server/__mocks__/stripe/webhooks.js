export default {
  constructEvent (body) {
    return body;
  }
};
