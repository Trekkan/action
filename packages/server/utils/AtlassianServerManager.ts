import fetch from 'node-fetch'
import makeAppLink from './makeAppLink'
import AtlassianManager from 'parabol-client/utils/AtlassianManager'

interface AuthQueryParams {
  grant_type: 'authorization_code'
  code: string
  redirect_uri: string
}

interface RefreshQueryParams {
  grant_type: 'refresh_token'
  refresh_token: string
}

interface OAuth2Response {
  access_token: string
  refresh_token: string
  error: any
  scope: string
}

class AtlassianServerManager extends AtlassianManager {
  fetch = fetch
  static async init(code: string) {
    return AtlassianServerManager.fetchToken({
      grant_type: 'authorization_code',
      code,
      redirect_uri: makeAppLink('auth/atlassian')
    })
  }

  static async refresh(refreshToken: string) {
    return AtlassianServerManager.fetchToken({
      grant_type: 'refresh_token',
      refresh_token: refreshToken
    })
  }

  static async fetchToken(partialQueryParams: AuthQueryParams | RefreshQueryParams) {
    const queryParams = {
      ...partialQueryParams,
      client_id: process.env.ATLASSIAN_CLIENT_ID,
      client_secret: process.env.ATLASSIAN_CLIENT_SECRET
    }

    const uri = `https://auth.atlassian.com/oauth/token`

    const tokenRes = await fetch(uri, {
      method: 'POST',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json'
      },
      body: JSON.stringify(queryParams)
    })

    const tokenJson = (await tokenRes.json()) as OAuth2Response

    const {access_token: accessToken, refresh_token: refreshToken, error, scope} = tokenJson
    if (error) {
      throw new Error(`Atlassian: ${error}`)
    }
    const providedScope = scope.split(' ')
    const matchingScope =
      new Set([...AtlassianServerManager.SCOPE.split(' '), ...providedScope]).size ===
      providedScope.length
    if (!matchingScope) {
      throw new Error(`bad scope: ${scope}`)
    }
    return new AtlassianServerManager(accessToken, refreshToken)
  }

  refreshToken?: string

  constructor(accessToken: string, refreshToken?: string) {
    super(accessToken)
    this.refreshToken = refreshToken
  }
}

export default AtlassianServerManager
