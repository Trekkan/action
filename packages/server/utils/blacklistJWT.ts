import getBlacklistJWTKey from './getBlacklistJWTKey'
import {Threshold, SubscriptionChannel} from 'parabol-client/types/constEnums'
import {toEpochSeconds} from './epochTime'
import publish from './publish'
import getRedis from './getRedis'

const blacklistJWT = async (userId: string, iat: number, mutatorId?: string) => {
  const key = getBlacklistJWTKey(userId)
  const redis = getRedis()
  const expiresIn = iat - toEpochSeconds(new Date()) + Math.floor(Threshold.JWT_LIFESPAN / 1000)
  await redis.set(key, iat, 'EX', expiresIn)
  publish(SubscriptionChannel.NOTIFICATION, userId, 'InvalidateSessionsPayload', {}, {mutatorId})
}

export default blacklistJWT
