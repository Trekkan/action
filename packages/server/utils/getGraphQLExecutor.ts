import {ExecutionResult} from 'graphql'
import {ServerChannel} from 'parabol-client/types/constEnums'
import PubSubPromise from './PubSubPromise'

export interface PubSubPromiseResponse extends ExecutionResult {
  jobId: string
}

let pubsub: PubSubPromise<PubSubPromiseResponse>
const getGraphQLExecutor = () => {
  if (!pubsub) {
    pubsub = new PubSubPromise(
      ServerChannel.GQL_EXECUTOR_REQUEST,
      ServerChannel.GQL_EXECUTOR_RESPONSE
    )
  }
  return pubsub
}
export default getGraphQLExecutor
