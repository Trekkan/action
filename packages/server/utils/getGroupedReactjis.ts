import Reactji from '../database/types/Reactji'
import {IReactji} from 'parabol-client/types/graphql'

const getGroupedReactjis = (reactjis: Reactji[], viewerId: string, idPrefix: string) => {
  const agg = {}
  for (let i = 0; i < reactjis.length; i++) {
    const {id, userId} = reactjis[i]
    const guid = `${idPrefix}:${id}`
    const isViewerReactji = viewerId === userId
    const record = agg[guid]
    if (!record) {
      agg[guid] = {id: guid, count: 1, isViewerReactji}
    } else {
      record.count++
      record.isViewerReactji = record.isViewerReactji || isViewerReactji
    }
  }
  return Object.values(agg) as IReactji[]
}

export default getGroupedReactjis
