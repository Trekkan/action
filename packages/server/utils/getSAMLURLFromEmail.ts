import base64url from 'base64url'
import getSSODomainFromEmail from 'parabol-client/utils/getSSODomainFromEmail'
import getRethink from '../database/rethinkDriver'
import {URL} from 'url'

const urlWithRelayState = (url: string, isInvited?: boolean | null) => {
  if (!isInvited) return url
  const relayState = base64url.encode(JSON.stringify({isInvited: true}))
  const urlObj = new URL(url)
  urlObj.searchParams.append('RelayState', relayState)
  return urlObj.toString()
}

const getSAMLURLFromEmail = async (email: string, isInvited?: boolean | null) => {
  const domainName = getSSODomainFromEmail(email)
  if (!domainName) return null
  const r = await getRethink()
  const baseURL = (await r
    .table('SAML')
    .getAll(domainName, {index: 'domain'})
    .nth(0)('url')
    .default(null)
    .run()) as string | null
  if (!baseURL) return null
  return urlWithRelayState(baseURL, isInvited)
}

export default getSAMLURLFromEmail
