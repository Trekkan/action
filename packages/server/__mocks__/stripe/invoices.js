import shortid from 'shortid';
import {deletedReturnVal, getDoc, updateFromOptions} from './utils';
import creditCardByToken from 'server/__tests__/utils/creditCardByToken';
import {toEpochSeconds} from 'server/utils/epochTime';
import {makeSubscriptionPlan} from './subscriptions';

export const makeInvoiceLine = (overrides) => {
  const now = new Date();
  const nowInSeconds = toEpochSeconds(now);
  const {id, metadata, periodStart, subscription} = overrides;
  return {
    "id": id,
    "object": "line_item",
    "amount": 0,
    "currency": "usd",
    "description": null,
    "discountable": true,
    "livemode": true,
    "metadata": metadata,
    "period": {
      "start": periodStart || 1488909033,
      "end": 1491501033
    },
    "plan": makeSubscriptionPlan(nowInSeconds),
    "proration": false,
    "quantity": 1,
    "subscription": subscription,
    "subscription_item": `si_${shortid.generate()}`,
    "type": "subscription"
  }
};

const makeNextMonthChargesLine = (subscription) => makeInvoiceLine({
  id: subscription.id,
  metadata: {
    orgId: subscription.metadata.orgId
  },
});

const makeInvoiceLines = (id, subscription, lineTypes) => {
  const totalCount = lineTypes ? Object.keys(lineTypes).reduce((count, type) => count + lineTypes[type], 0) : undefined;
  return {
    "data": [
      makeNextMonthChargesLine(subscription),
    ],
    "total_count": totalCount,
    "object": "list",
    "url": `/v1/invoices/${id}/lines`
  }
};

export const createNewInvoice = (overrides, reject) => {
  const {id, customer, subscription} = overrides;
  return {
    "id": id,
    "object": "invoice",
    "amount_due": 0,
    "application_fee": null,
    "attempt_count": 0,
    "attempted": true,
    "charge": null,
    "closed": true,
    "currency": "usd",
    "customer": customer,
    "date": toEpochSeconds(new Date()),
    "description": null,
    "discount": null,
    "ending_balance": 0,
    "forgiven": false,
    "lines": makeInvoiceLines(id, subscription),
    "livemode": false,
    "metadata": {},
    "next_payment_attempt": null,
    "paid": true,
    "period_end": 1488909033,
    "period_start": 1488909033,
    "receipt_number": null,
    "starting_balance": 0,
    "statement_descriptor": null,
    "subscription": subscription.id,
    "subtotal": 0,
    "tax": null,
    "tax_percent": null,
    "total": 0,
    "webhooks_delivered_at": 1488909040
  }
};

export default (stripe) => ({
   create: jest.fn((options) => new Promise((resolve, reject) => {
     const {id} = options;
     const invoiceDoc = stripe.__db.invoices[id] = createNewInvoice(options, reject);
     resolve(invoiceDoc);
   })),
  retrieve: jest.fn((id) => new Promise((resolve, reject) => {
    const doc = getDoc(stripe.__db.invoices[id], reject);
    resolve(doc);
  })),
  retrieveLines: jest.fn((id) => new Promise((resolve, reject) => {
    const doc = getDoc(stripe.__db.invoices[id], reject);
    resolve(doc.lines);
  })),
  //TODO
  retrieveUpcoming: jest.fn((id) => new Promise((resolve, reject) => {
    const doc = getDoc(stripe.__db.invoices[id], reject);
    resolve(doc);
  })),
  //
  update: jest.fn((id, options) => new Promise((resolve, reject) => {
    const doc = getDoc(stripe.__db.invoices[id], reject);
    updateFromOptions(doc, stripe.invoices.__updateHandlers, options, reject);
    resolve(doc)
  })),
  del: jest.fn((id) => new Promise((resolve, reject) => {
    getDoc(stripe.__db.invoices[id], reject);
    delete stripe.__db.invoices[id];
    resolve(deletedReturnVal(id));
  })),
  __trimFields: [
    'id',
    'customer',
    'metadata.orgId',
    'subscription',
    'lines.url',
    'lines.data.id',
    'lines.data.metadata.orgId',
    'lines.data.subscription_item'
  ],
  __triggers: ['update', 'del'],
  __uniqueKeyField: 'customer',
  __updateHandlers: {}
});
