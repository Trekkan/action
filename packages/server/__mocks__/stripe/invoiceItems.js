import {getDoc, updateFromOptions} from './utils';
import {makeInvoiceLine} from './invoices';

export default (stripe) => ({
  create: jest.fn((options) => new Promise((resolve) => {
    const {id} = options;
    const invoiceDoc = stripe.__db.invoiceItems[id] = makeInvoiceLine(options);
    resolve(invoiceDoc);
  })),
  retrieve: jest.fn((id) => new Promise((resolve, reject) => {
    const doc = getDoc(stripe.__db.invoiceItems[id], reject);
    resolve(doc);
  })),
  update: jest.fn((id, options) => new Promise((resolve, reject) => {
    const doc = getDoc(stripe.__db.invoiceItems[id], reject);
    updateFromOptions(doc, stripe.invoiceItems.__updateHandlers, options, reject);
    resolve(doc)
  })),
  __trimFields: [
    'id',
    'customer',
    'metadata.userId',
    'subscription',
    'subscription_item'
  ],
  __triggers: ['update'],
  __uniqueKeyField: 'subscription',
  __updateHandlers: {}
});