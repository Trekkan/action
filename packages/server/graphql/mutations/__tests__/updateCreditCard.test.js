/* eslint-env jest */
import DynamicSerializer from 'dynamic-serializer'
import MockDate from 'mockdate'
import makeDataLoader from '../../../__tests__/setup/makeDataLoader'
import mockAuthToken from '../../../__tests__/setup/mockAuthToken'
import MockDB from '../../../__tests__/setup/MockDB'
import { __now } from '../../../__tests__/setup/mockTimes'
import fetchAndSerialize from '../../../__tests__/utils/fetchAndSerialize'
import getRethink from '../../../database/rethinkDriver'
import { PRO } from 'parabol-client/utils/constants'
import updateCreditCard from '../updateCreditCard'
import stripe from '../../../billing/stripe'

MockDate.set(__now)
console.error = jest.fn()

describe('UpdateCreditCard', () => {
  test('can call updateCreditCard without a stripe subscription', async () => {
    // SETUP
    const r = await getRethink()
    const dynamicSerializer = new DynamicSerializer()
    const mockDB = new MockDB()
    const { organization, user } = await mockDB
      .init({ plan: PRO })
      .organization(0, { stripeSubscriptionId: null })
    const org = organization[0]
    const { id: orgId } = org
    stripe.__setMockData(org)
    const authToken = mockAuthToken(user[0])
    const dataLoader = makeDataLoader(authToken)

    // TEST
    await updateCreditCard.resolve(
      undefined,
      { orgId, stripeToken: 'tok_4242424242424242' },
      { authToken, dataLoader }
    )

    // VERIFY
    const db = await fetchAndSerialize(
      {
        organization: r.table('Organization').get(orgId),
        team: r.table('Team').getAll(orgId, { index: 'orgId' })
      },
      dynamicSerializer
    )
    expect(db).toMatchSnapshot()
    expect(stripe.__snapshot(org.stripeId, dynamicSerializer)).toMatchSnapshot()
  })
})
