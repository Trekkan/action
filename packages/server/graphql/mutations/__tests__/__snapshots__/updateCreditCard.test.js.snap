// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UpdateCreditCard can call updateCreditCard without a stripe subscription 1`] = `
Object {
  "organization": Object {
    "createdAt": 2017-01-01T07:00:00.000Z,
    "creditCard": Object {
      "brand": "Visa",
      "expiry": "12/23",
      "id": "card_124",
      "last4": "4242",
      "number": "4012888888881881",
    },
    "id": 0,
    "name": "The Averagers, Inc.",
    "orgUsers": Array [
      Object {
        "id": 2,
        "inactive": false,
        "role": "billingLeader",
      },
      Object {
        "id": 3,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 4,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 5,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 6,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 7,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 8,
        "inactive": false,
        "role": null,
      },
      Object {
        "id": 9,
        "inactive": false,
        "role": null,
      },
    ],
    "periodStart": 2017-01-01T07:00:00.000Z,
    "stripeId": 1,
    "stripeSubscriptionId": null,
    "updatedAt": 2017-01-01T08:00:00.000Z,
  },
  "team": Array [
    Object {
      "activeFacilitator": null,
      "facilitatorPhase": "lobby",
      "facilitatorPhaseItem": null,
      "id": 10,
      "isArchived": false,
      "isPaid": true,
      "meetingId": null,
      "meetingPhase": "lobby",
      "meetingPhaseItem": null,
      "name": "Team America",
      "orgId": 0,
      "tier": "personal",
      "updatedAt": 2017-01-01T08:00:00.000Z,
    },
  ],
}
`;

exports[`UpdateCreditCard can call updateCreditCard without a stripe subscription 2`] = `
Object {
  "customers": Array [
    Object {
      "account_balance": 0,
      "created": 1483257600,
      "currency": "usd",
      "default_source": "card_123",
      "delinquent": false,
      "description": null,
      "discount": null,
      "email": null,
      "id": 1,
      "livemode": false,
      "metadata": Object {
        "orgId": 0,
      },
      "object": "customer",
      "shipping": null,
      "sources": Object {
        "data": Array [
          Object {
            "address_city": null,
            "address_country": null,
            "address_line1": null,
            "address_line1_check": null,
            "address_line2": null,
            "address_state": null,
            "address_zip": null,
            "address_zip_check": null,
            "brand": "Visa",
            "country": "US",
            "customer": 1,
            "cvc_check": "pass",
            "dynamic_last4": null,
            "exp_month": "12",
            "exp_year": "2023",
            "funding": "credit",
            "id": "card_123",
            "last4": "4242",
            "metadata": Object {},
            "name": null,
            "object": "card",
            "tokenization_method": null,
          },
        ],
        "has_more": false,
        "object": "list",
        "total_count": 1,
        "url": 11,
      },
    },
  ],
}
`;
